README
aidentd - The Afghan Ident Daemon (README based on original for nullidentd)
--------------------------------------------------------------------------

Version 1.0
Copyright 2001 Afghan Hound <afghan-ident@afghanhound.org.uk>

First public release - No further changes are likely except for code
simplification, portability fixes or adding new Pokemon.

What
----

aidentd is intended to be a total joke of an ident server, based on
the much cleaner ( and better) nullident by Brian Young <bayoung@acm.org>.

The program implements the auth protocol from RFC 1413.  This protocol is
used to identify active TCP connections.  It returns a fake response for 
any request - consisting of a random OS from RFC1340 and a random
userid from the (current) list of active Pokemon. 

The auth port is tcp 113.

See the original nullidentd readme (README.null) for philosophical
issues and slightly more detail.


Goals
-----

It seemed like a good idea at the time. But many things do after lots
of whisky. It's probably fortunate that RFC1413 doesn't allow carriage
returns, or aidentd would probably be returning ASCII art Snoopy's.

It might just cheer up some depressed mail/ftp/etc server admin perusing 
their logs, or it might drive them over the edge: who knows?

Installation (modified from nullidentd)
---------------------------------------

Build the program by simply typing 'make'.  Edit the makefile if it's
required.  There is a 'make install' rule, but you may want to edit the
makefile to set the installation directory.

aidentd has been built and tested on a Linux 2.2 system.  Portability
fixes are welcomed.

aidentd must be run from inetd.  The following line in inetd.conf should
do the job:

auth   stream  tcp     nowait    nobody    /usr/local/sbin/aidentd aidentd

 - or -

auth   stream  tcp     nowait    nobody   /usr/sbin/tcpd aidentd

if you've installed aidentd in /usr/sbin and want to use
hosts.allow/deny.

Gotta Catch 'Em All
-------------------

Pokemon is (are?) a trademark of Nintendo. The various OS names are 
probably trademarks of whoever owns them.

Afghan Hound : 	afghan@afghanhound.org.uk
               	http://www.afghanhound.org.uk/
		http://www.uklinux.afghanhound.net/ - for the office
		Quake statistics page.
	
                  ***Afghan Hound Software***
	
                      M" " " " " " " " "o"
                     oo      "       " "o
                     "o" " "   " " " " M
              o" " " " " " " " " " " " "
             oo" " " "o"   "   " "o" "oo
            o" " "   "o" " " " " "   " "o o
           o     " "  o"  o" " "  oM " " "oMo
          M" " "oMoMo  "o" " " " " "oM  o"o"
          o"  oMo"oMo"o"oMoMoM "   " "o" "oM
          o" "oMo" "oMo"MMoMMMo"o"o"o"o" "o"
         Mo"  oMoMo"oM"oo""" "oMo"o"o"o"o oM
         M "oMoM"  "     Mo oM Mo"o"o"o"o"o""
         Mo"oMo"           ""  Mo"o"oMoMo"oM
         MoM                    o"o"oMMMo"o"
       o ""                     o"o" ""Mo"oM
    oM                          o"o"oMM"o"o"
    MM              oooMM" o    o"o"o"MMo"o"
    "M          ooo"""     MM    "o"o"o"o"o"M
     "Mo  oooMMM           M     "o" "o"o"o"o
      ""MMMMoMoM                 " "o"o"o"o"o
          MMMMM"                 "oM "MMo"o"o
          oMMM"                  MoMoMoMoMoM"
           "                         "

          ***Delivering pointless software since 1999***





CHANGELOG
aidentd
----------

Copyright 2001 Afghan Hound <afghan-ident@afghanhound.org.uk>

Version 1.0
-----------

 - Initial, (and probably final) release.



Original nullidentd changelog:


nullidentd
----------

Copyright 1999 Brian Young <bayoung@acm.org>

Version 1.0
-----------

- no code changes
- updated README

Version 0.9.1
-------------

- send \r\n instead of \n (as per RFC1413)
- use snprintf instead of sprintf
    - no security hole, but it's just wrong to use sprintf in a daemon
- use a session timeout instead of idle timeouts
    - makes more sense for such a small protocol

Version 0.9
-----------

- first public release



aidentd.c
/*  aidentd - a fake ident server

    Copyright (C) 2000 Afghan Hound Software

    Based on:
    
    nullidentd.c
    Copyright 1999 Brian Young <bayoung@acm.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA*/

/* aidentd.c - the server itself */

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/time.h>
#include <unistd.h>
#include <signal.h>
#include "fake.h"
#include "version.h"

/* In seconds */
#define SESSION_TIMEOUT 20

#define MAX_RESPONSE 200
#define MAX_REQUEST  100
#define MAX_USERID   50

int get_random(int max)
{
  return (int)((1.0*rand()*max/(RAND_MAX+1.0)));
}

void init_random()
{
  struct timeval tv;
  long   wibble;
  gettimeofday(&tv,(struct timezone *)NULL);
  wibble = (int)tv.tv_sec + (int)tv.tv_usec;
  srand( (int)wibble );
} 

int write_response( int fd, char *response, int len )
{
  int		retval;
  int		byteswritten = 0;

  while( byteswritten < len ) 
    {
      retval = write( fd, response + byteswritten, len - byteswritten );
      if( retval <= 0 ) 
	{
	  /* error */
	  return 0;
	}
      byteswritten += retval;
    }

  /* success */
  return 1;
}

int read_request( int fd, char *request, int maxlen )
{
  int		retval;
  char	c;
  int		bytesread = 0;

  /* read until \n */
  while( bytesread < maxlen - 1 ) 
    {
      if( read( fd, &c, 1 ) != 1 ) 
	{
	  /* error */
	  return 0;
	}
      if( c == '\n' ) 
	{
	  /* end of line */
	  break;
	}
      
      request[bytesread] = c;
      bytesread += 1;
    }
  
  if( bytesread > 0 && request[bytesread-1] == '\r' ) 
    {
      /* strip trailing \r */
      bytesread -= 1;
    }
  /* null terminate */
  request[bytesread] = '\0';

  /* success */
  return 1;
}

void session_timeout( int foo )
{
  exit( 0 );
}

int main( int argc, const char *argv[] )
{
  char		c;
  int		infd;
  int		outfd;
  int		response_len;
  char		response[MAX_RESPONSE];
  char		request[MAX_REQUEST];

  if( ( getgid() == 0 ) || ( getuid() == 0 )  ) 
    {
      fprintf( stderr, "Group or user id is root, exitting.\n" );
      return 1;
    }


  infd = fileno( stdin );
  outfd = fileno( stdout );
  init_random();

  /* set the session timeout */
  signal( SIGALRM, session_timeout );
  alarm( SESSION_TIMEOUT );

  for( ;; ) 
    {
      /* read the request */
      if( !read_request( infd, request, MAX_REQUEST ) ) 
	{
	  /* error or timed out */
	  goto done;
	}

    /* format the response */
      response_len = snprintf( response, 
			       sizeof( response ), 
			       "%.20s : USERID : %s : %s\r\n", 
			       request, 
			       os_types[get_random(NUM_OS)],
			       userids[get_random(NUM_USER)]);

      /* send the line */
      if( !write_response( outfd, response, response_len ) ) 
	{
	  goto done;
	}
    }
  
 done:
  return 0;
}




fake.h
/*  aidentd - a fake ident server

    Copyright (C) 2000 Afghan Hound Software

    Based on:    
    nullidentd.c
    Copyright 1999 Brian Young <bayoung@acm.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/* fake.h - OS's and user names    */
/* For extra fun, swap over the user_ids/os_types and 
   NUM_OS/NUM_USER bits. */

#define NUM_OS   118
#define NUM_USER 153

/* From RFC1340 - extra ones added just for the sake of it.   */

char *os_types[NUM_OS]={
	"AEGIS","LISP","SUN_OS_3.5","APOLLO",
	"LISPM","SUN_OS_4.0","AIX/370","LOCUS",
	"SWIFT","AIX-PS/2","MACOS","TAC","BS-2000",
	"MINOS","TANDEM","CEDAR","MOS","TENEX","CGW",
	"MPE5","TOPS10","CHORUS","MSDOS","TOPS20",
	"CHRYSALIS","MULTICS","TOS","CMOS","MUSIC",
	"TP3010","CMS","MUSIC/SP","TRSDOS","COS","MVS",
	"ULTRIX","CPIX","MVS/SP","UNIX","CTOS","NEXUS",
	"UNIX-BSD","CTSS","NMS","UNIX-V1AT","DCN","NONSTOP",
	"UNIX-V","DDNOS","NOS-2","UNIX-V.1","DOMAIN","NTOS",
	"UNIX-V.2","DOS","OS/DDP","UNIX-V.3","EDX","OS/2",
	"UNIX-PC","ELF","OS4","UNKNOWN","EMBOS","OS86",
	"UT2D","EMMOS","OSX","V","EPOS","PCDOS","VM","FOONEX",
	"PERQ/OS","VM/370","FUZZ","PLI","VM/CMS","GCOS","PSDOS/MIT",
	"VM/SP","GPOS","PRIMOS","VMS","HDOS","RMX/RDOS",
	"VMS/EUNICE","IMAGEN","ROS","VRTX","INTERCOM",
	"RSX11M","WAITS","IMPRESS","RTE-A","WANG","INTERLISP",
	"SATOPS","WIN32","IOS","SCO-XENIX/386","X11R3","IRIX",
	"SCS","XDE","ISI-68020","SIMP","XENIX","ITS","SUN",
	"FREEBSD","OPENBSD","NETBSD","LINUX","QNX","WINNT","MINIX",
	"SOLARIS"
	};

/* From watching the cartoons. */

char *userids[NUM_USER] = {
	"Bulbasaur","Ivysaur","Venusaur","Charmander","Charmeleon",
	"Charizard","Squirtle","Wartortle","Blastoise","Caterpie",
	"Metapod","Butterfree","Weedle","Kakuna","Beedrill",
	"Pidgey","Pidgeotto","Pidgeot","Rattata","Raticate",
	"Spearow","Fearow","Ekans","Arbok",
/* Pi...Pi...Pi....       */
	"Pikachu",
/* uuuuuuuuuuu!!!!!!!!!!! */
	"Raichu","Sandshrew","Sandslash","Nidoran_girl","Nidorina",
	"Nidoqueen","Nidoran_boy","Nidorino","Nidoking",
	"Clefairy","Clefable","Vulpix","Ninetales","Jigglypuff",
	"Wigglytuff","Zubat","Golbat","Oddish","Gloom","Vileplume",
	"Paras","Parasect","Venonat","Venomoth","Diglett","Dugtrio",
	"Meowth","Persian","Psyduck","Golduck","Mankey","Primeape",
	"Growlithe","Arcanine","Poliwag","Poliwhirl","Poliwrath",
	"Abra","Kadabra","Alakazam","Machop","Machoke","Machamp",
	"Bellsprout","Weepinbell","Victreebell","Tentacool",
	"Tentacruel","Geodude","Graveler","Golem","Ponyta",
	"Rapidash","Slowpoke","Slowbro","Magnemite","Magneton",
	"Farfetch'd","Doduo","Dodrio","Seel","Dewgong","Grimer",
	"Muk","Shellder","Cloyster","Gastly","Haunter","Gengar",
	"Onix","Drowzee","Hypno","Krabby","Kingler","Voltorb",
	"Electrode","Exeggcute","Exeggutor","Cubone","Marowak",
	"Hitmonlee","Hitmonchan","Lickitung","Koffing","Weezing",
	"Rhyhorn","Rhydon","Chansey","Tangela","Kangaskan",
	"Horsea","Seadra","Goldeen","Seaking","Staryu","Starmie",
	"Mr_Mime","Scyther","Jynx","Electabuzz","Magmar","Pinsir",
	"Tauros","Magikarp","Gyarados","Lapras","Ditto","Eevee",
	"Vaporeon","Jolteon","Flareon","Porygon","Omanite","Omastar",
	"Kabuto","Kabutops","Aerodactyl","Snorlax","Articuno",
	"Zapdos","Moltres","Dratini","Dragonair","Dragonite",
	"Mewtwo","Mew","Houou","Togepi"
};
